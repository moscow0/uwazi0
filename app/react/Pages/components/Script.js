import { actions } from 'app/BasicReducer';

import { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

class Script extends Component {
  constructor(props) {
    super(props);

    this.scriptElement = null;
  }

  componentDidMount() {
    this.appendScript();
    window.addEventListener('error', this.props.onError);
    window.addEventListener('unhandledrejection', this.props.onError);
  }

  componentDidUpdate(prevProps) {
    const { children } = this.props;
    if (children !== prevProps.children) {
      this.removeScript();
      window.removeEventListener('error', this.props.onError);
      window.removeEventListener('unhandledrejection', this.props.onError);
      this.appendScript();
    } else if (this.scriptElement === null) {
      this.appendScript();
    }
  }

  componentWillUnmount() {
    this.removeScript();
    window.removeEventListener('error', this.props.onError);
    window.removeEventListener('unhandledrejection', this.props.onError);
  }

  appendScript() {
    const { children, scriptRendered } = this.props;
    if (children && scriptRendered === false) {
      const s = document.createElement('script');
      s.src = `data:text/javascript,(function(){${encodeURIComponent(`\n\n${children}\n\n`)}})()`;
      document.body.appendChild(s);
      this.scriptElement = s;
      this.props.dispatch(actions.setIn('page/pageView', 'scriptRendered', true));
    }
  }

  removeScript() {
    if (this.scriptElement) {
      this.scriptElement.remove();
      this.scriptElement = null;
    }
  }

  render() {
    return null;
  }
}

Script.defaultProps = {
  children: '',
  scriptRendered: null,
  onError: () => {},
};

Script.propTypes = {
  children: PropTypes.string,
  scriptRendered: PropTypes.bool,
  dispatch: PropTypes.func.isRequired,
  onError: PropTypes.func,
};

const container = connect()(Script);

export default container;
