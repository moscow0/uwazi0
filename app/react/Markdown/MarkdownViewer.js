import PropTypes from 'prop-types';
import React, { Component } from 'react';
import rison from 'rison-node';
import { Translate } from 'app/I18N';
import { MarkdownLink, SearchBox, MarkdownMedia, ItemList } from './components';
import CustomHookComponents from './CustomHooks';

import markdownToReact from './markdownToReact';

class MarkdownViewer extends Component {
  static errorHtml(index, message) {
    return (
      <p key={index} className="error">
        <br />
        <strong>
          <i>
            <Translate translationKey="Custom component error">
              Custom component markup error: unsuported values! Please check your configuration
            </Translate>
          </i>
        </strong>
        <br />
        {message}
        <br />
      </p>
    );
  }

  static customHook(config, index) {
    const props = rison.decode(config);
    if (!CustomHookComponents[props.component]) {
      throw new Error('Invalid  component');
    }
    const Element = CustomHookComponents[props.component];
    return <Element {...props} key={index} />;
  }

  inlineComponent(type, config, index) {
    const { compact } = this.props;
    let result;
    if (type === 'list') {
      result = this.list(config, index);
    }

    if (type === 'link') {
      result = <MarkdownLink {...rison.decode(config)} key={index} />;
    }

    if (type === 'searchbox') {
      result = <SearchBox {...rison.decode(config)} key={index} />;
    }

    if (['vimeo', 'youtube', 'media'].includes(type)) {
      result = (
        <div key={index}>
          <MarkdownMedia key={index} config={config} compact={compact} />
        </div>
      );
    }

    if (type === 'customhook') {
      result = MarkdownViewer.customHook(config, index);
    }
    return result;
  }

  customComponent(type, config, index, children) {
    try {
      if (typeof type === 'function' || typeof type === 'object') {
        const Element = type;
        return (
          <Element {...config} key={index}>
            {children}
          </Element>
        );
      }

      if (type) {
        return this.inlineComponent(type, config, index);
      }
    } catch (error) {
      return MarkdownViewer.errorHtml(index, error.message);
    }

    return false;
  }

  list(_config, index) {
    const listData = this.props.lists[this.renderedLists] || {};
    const output = (
      <ItemList
        key={index}
        link={`/library/${listData.params}`}
        items={listData.items}
        options={listData.options}
      />
    );
    this.renderedLists += 1;
    return output;
  }

  render() {
    this.renderedLists = 0;

    const ReactFromMarkdown = markdownToReact(
      this.props.markdown,
      this.customComponent.bind(this),
      this.props.html
    );

    if (!ReactFromMarkdown) {
      return false;
    }

    return <div className="markdown-viewer">{ReactFromMarkdown}</div>;
  }
}

MarkdownViewer.defaultProps = {
  lists: [],
  markdown: '',
  html: false,
  compact: false,
};

MarkdownViewer.propTypes = {
  markdown: PropTypes.string,
  lists: PropTypes.arrayOf(PropTypes.object),
  html: PropTypes.bool,
  compact: PropTypes.bool,
};

export default MarkdownViewer;
