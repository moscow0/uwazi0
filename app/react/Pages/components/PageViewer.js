import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import Footer from 'app/App/Footer';
import MarkdownViewer from 'app/Markdown';
import { Context } from 'app/Markdown/components';
import { ErrorBoundary } from 'app/App/ErrorHandling/ErrorBoundary';
import { Icon } from 'UI';
import { Translate } from 'app/I18N';
import { ErrorFallback } from 'app/App/ErrorHandling/ErrorFallback';
import { parseRenderingError } from 'app/App/ErrorHandling/ErrorUtils';
import Script from './Script';

const parseSSRError = error => {
  const SSRError = error instanceof Immutable.Map ? error.toJS() : error;
  return SSRError?.json ? parseRenderingError(SSRError) : null;
};

class PageViewer extends Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.warningPageError = this.warningPageError.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.page.get('title') !== this.props.page.get('title')) {
      this.removeCustomPageError();
    }
  }

  componentWillUnmount() {
    if (this.state.customPageError) {
      this.removeCustomPageError();
    }
  }

  warningPageError(error) {
    this.setState({ customPageError: error });
  }

  removeCustomPageError() {
    this.setState({ customPageError: null });
  }

  renderErrorWarning() {
    return (
      <div className="alert alert-danger">
        <Icon icon="exclamation-triangle" />
        <Translate>
          There is an unexpected error on this custom page, it may not work properly. Please contact
          an admin for details.
        </Translate>
        <Icon icon="times" onClick={() => this.removeCustomPageError()} />
      </div>
    );
  }

  render() {
    const { page, itemLists, datasets, error, setBrowserTitle } = this.props;
    const lists = itemLists.toJS();
    const originalText = page.getIn(['metadata', 'content']) || '';
    const scriptRendered = page.getIn(['scriptRendered']);
    let scriptCode = page.getIn(['metadata', 'script']) || '';
    scriptCode = `var datasets = window.store.getState().page.datasets.toJS();
    ${scriptCode}`;
    const parsedPageError = parseSSRError(error);
    return (
      <div className="row">
        {!parsedPageError && (
          <>
            {setBrowserTitle && (
              <Helmet>
                <title>{page.get('title') ? page.get('title') : 'Page'}</title>
              </Helmet>
            )}
            <main className="page-viewer document-viewer">
              <div className="main-wrapper">
                {this.state.customPageError && this.renderErrorWarning()}
                <Context.Provider value={datasets}>
                  <ErrorBoundary>
                    <MarkdownViewer html markdown={originalText} lists={lists} />
                  </ErrorBoundary>
                </Context.Provider>
                <Footer />
              </div>
            </main>
            <Script scriptRendered={scriptRendered} onError={e => this.warningPageError(e)}>
              {scriptCode}
            </Script>
          </>
        )}
        {parsedPageError && (
          <div className="main-wrapper">
            <ErrorFallback error={parsedPageError} />
            <Footer />
          </div>
        )}
      </div>
    );
  }
}

PageViewer.defaultProps = {
  page: Immutable.fromJS({}),
  itemLists: Immutable.fromJS([]),
  datasets: Immutable.fromJS({}),
  error: Immutable.fromJS({}),
  setBrowserTitle: true,
};

PageViewer.propTypes = {
  page: PropTypes.instanceOf(Immutable.Map),
  itemLists: PropTypes.instanceOf(Immutable.List),
  datasets: PropTypes.instanceOf(Immutable.Map),
  error: PropTypes.instanceOf(Immutable.Map),
  setBrowserTitle: PropTypes.bool,
};

const mapStateToProps = ({ page }) => ({
  page: page.pageView,
  datasets: page.datasets,
  itemLists: page.itemLists,
  error: page.error,
});

const container = connect(mapStateToProps)(PageViewer);

export { container as PageViewer };
