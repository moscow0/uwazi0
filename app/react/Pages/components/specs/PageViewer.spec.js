import Immutable from 'immutable';
import React from 'react';
import { Helmet } from 'react-helmet';
import { shallow } from 'enzyme';

import MarkdownViewer from 'app/Markdown';

import { PageViewer } from '../PageViewer';
import Script from '../Script';

describe('PageViewer', () => {
  let component;
  let props;
  let context;

  beforeEach(() => {
    props = {
      page: Immutable.fromJS({
        _id: 1,
        title: 'Page 1',
        metadata: /*non-metadata-object*/ { content: 'MarkdownContent', script: 'JSScript' },
      }),
      itemLists: Immutable.fromJS([{ item: 'item' }]),
      datasets: Immutable.fromJS({ key: 'value' }),
    };
  });

  const render = () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    component = shallow(<PageViewer.WrappedComponent {...props} />, { context });
  };

  describe('render', () => {
    beforeEach(() => {
      render();
    });

    it('should render a MarkdownViewer with the markdown and the items for the lists', () => {
      expect(component.find(MarkdownViewer).props().markdown).toBe('MarkdownContent');
      expect(component.find(MarkdownViewer).props().lists).toEqual([{ item: 'item' }]);
    });

    it('should render the script', () => {
      const scriptElement = component.find(Script);
      expect(scriptElement).toMatchSnapshot();
    });

    describe('Helmet', () => {
      it('should render the page helmet', () => {
        expect(component.find(Helmet).find('title').text()).toBe('Page 1');
      });

      it('should not overwrite the page title', () => {
        props.setBrowserTitle = false;
        render();
        expect(component.find(Helmet).length).toBe(0);
      });
    });
  });
});
