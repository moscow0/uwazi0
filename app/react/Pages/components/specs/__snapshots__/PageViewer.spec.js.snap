// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageViewer render should render the script 1`] = `
<Connect(Script)
  onError={[Function]}
>
  var datasets = window.store.getState().page.datasets.toJS();
    JSScript
</Connect(Script)>
`;
